data/src/main/java/org/apache/iceberg/data/BaseDeleteLoader.java

```diff
@@ -0,0 +1,260 @@
+    this(loadInputFile, ThreadPools.getDeleteWorkerPool());
+  }
+
+  public BaseDeleteLoader(
+      Function<DeleteFile, InputFile> loadInputFile, ExecutorService workerPool) {
+    this.loadInputFile = loadInputFile;
+    this.workerPool = workerPool;
+  }
+
+  /**
+   * Checks if the given number of bytes can be cached.
+   *
+   * <p>Implementations should override this method if they support caching. It is also recommended
+   * to use the provided size as a guideline to decide whether the value is eligible for caching.
+   * For instance, it may be beneficial to discard values that are too large to optimize the cache
+   * performance and utilization.
+   */
+  protected boolean canCache(long size) {
+    return false;
+  }
+
+  /**
+   * Gets the cached value for the key or populates the cache with a new mapping.
+   *
+   * <p>If the value for the specified key is in the cache, it should be returned. If the value is
+   * not in the cache, implementations should compute the value using the provided supplier, cache
+   * it, and then return it.
+   *
+   * <p>This method will be called only if {@link #canCache(long)} returned true.
+   */
+  protected <V> V get(String key, Supplier<V> valueSupplier, long valueSize) {
+    throw new UnsupportedOperationException(getClass().getName() + " does not support caching");
+  }
+
+  @Override
+  public StructLikeSet loadEqualityDeletes(Iterable<DeleteFile> deleteFiles, Schema projection) {
+    Iterable<Iterable<StructLike>> deletes =
+        execute(deleteFiles, deleteFile -> getOrLoadEqDeletes(deleteFile, projection));
+    StructLikeSet deleteSet = StructLikeSet.create(projection.asStruct());
+    Iterables.addAll(deleteSet, Iterables.concat(deletes));
+    return deleteSet;
+  }
+
+  private Iterable<StructLike> getOrLoadEqDeletes(DeleteFile deleteFile, Schema projection) {
+    long estimatedSize = estimateEqDeletesSize(deleteFile, projection);
+    if (canCache(estimatedSize)) {
+      String cacheKey = deleteFile.path().toString();
+      return get(cacheKey, () -> loadEqDeletes(deleteFile, projection), estimatedSize);
+    } else {
+      return loadEqDeletes(deleteFile, projection);
+    }
+  }
+
+  private Iterable<StructLike> loadEqDeletes(DeleteFile deleteFile, Schema projection) {
+    CloseableIterable<Record> deletes = openDeletes(deleteFile, projection);
+    CloseableIterable<Record> copiedDeletes = CloseableIterable.transform(deletes, Record::copy);
+    CloseableIterable<StructLike> copiedDeletesAsStructs = toStructs(copiedDeletes, projection);
+    return materialize(copiedDeletesAsStructs);
+  }
+
+  private CloseableIterable<StructLike> toStructs(
+      CloseableIterable<Record> records, Schema schema) {
+    InternalRecordWrapper wrapper = new InternalRecordWrapper(schema.asStruct());
+    return CloseableIterable.transform(records, wrapper::copyFor);
+  }
+
+  private <T> Iterable<T> materialize(CloseableIterable<T> iterable) {
```

Comment:
This essentially materializes the iterable and closes the resources so that the actual value can be cached. We cannot cache the iterable directly cause it is lazy. Let me think.

Reply:
Added a comment for now. Still thinking.

Reply:
I'd probably stick with the comment.